### PR TITLE
Flags module update

### PR DESCRIPTION
**Pull request by PabloOQ**
#202 

Took longer than expected, as such I could not implement the groups feature, but I think it shouldn't take that much as I think all the logic needed is there.

I updated the strings to match the new functionality, I expect weblate to mark the translations as outdated, but I don't know how weblate works, just in case I warn about it here.

There is a search bar now, it detects the words in any order.

I created a new class to store flags in there, there is a TODO, I did not do it because it's unclear to me if it should be implemented or not. If somehow a non-compatible flag is received, should we store it? If that were to be the case we could use the hex value of the flag as its name.

Currently all flags preferences are stored, even if they are the default value, but if this is changed it should still work.
Because I was running out of time (I will be busy for some time), I couldn't implement something that I think is a must, a way to set all flags as automatic/off. I don't think a button to put everything to on, or shown, would make much sense
They are stored in a json with this format:

```
{
    "groups":
    {
        "default":
        {
            "FLAG_ACTIVITY_BROUGHT_TO_FRONT":
            {
                "state": 1,
                "show": true
            },
            "FLAG_ACTIVITY_CLEAR_TASK":
            {
                "state": 2,
                "show": true
            },
            "FLAG_ACTIVITY_CLEAR_TOP":
            {
                "state": 0,
                "show": true
            },
            "FLAG_ACTIVITY_EXCLUDE_FROM_RECENTS":
            {
                "state": 1,
                "show": false
            },
            "FLAG_ACTIVITY_FORWARD_RESULT":
            {
                "state": 2,
                "show": false
            },
            "FLAG_ACTIVITY_LAUNCHED_FROM_HISTORY":
            {
                "state": 0,
                "show": false
            },
            .
            .
            .
            "FLAG_INCLUDE_STOPPED_PACKAGES":
            {
                "state": 0,
                "show": false
            }
        }
    }
}
```

There is also this graphical bug, I don't know why it happens, probably some basic knowledge about android I just lack, hopefully you are able to fix it. At the bottom, with the other buttons, the arrow of the spinner is just,... there, doing something. If you fix this, please tag me, I want to know why it happens.

https://user-images.githubusercontent.com/55493443/230743927-370d2362-8379-4c93-825a-02c1ca61a8da.mp4

In the dialog there are 2 color indicators, the left one shows the value received by the intent, the right one shows the preference. I didn't want to use red and green, but its what I think felt better, previosly I had the colors change depending of the value, no color if the value was not changed according to the intent or the preferences, yellow if it was different, and grey if the preference was set on auto. It was very confusing.

I have only tested this on my phone as, I had to delete the emulators to free up some disk space and haven't cleaned it up yet. So older versions of android are still untested.

Let me know what you think!

### CODE REVIEW

**Comment by TrianguloY** [comment]
Lots to unpack here! Let's see if I can review and merge it this weekend (maybe next one though, I have a couple ongoing changes I also want to work on)

Thank you!

**Comment by TrianguloY** [comment]
Update! 

I've (finally) started reviewing the pr. Overall it's pretty stable (except for a crash on startup if the flags config file doesn't exists. Oops) but visually it's a bit confusing. I'm trying to change the layout to see if it can be improved somehow, but it's not easy (and don't worry, I'll keep those colored bars, I really like them!).

Hopefully this weekend I'll be able to approve it! Sorry for the delay.

**Comment by PabloOQ** [comment]
> except for a crash on startup if the flags config file doesn't exists

I thought I checked that, looks like I didn't. Maybe it's because I did some refactors at the end. Thanks for letting me now!

>  but visually it's a bit confusing

What do you mean?
If it's hard to follow because it's too much text, then alternating colors could be a solution. Which, I agree, it's a giant wall of text, I don't think there is a way around it.
image


> Sorry for the delay.

Don't worry!

**Comment by TrianguloY** [comment]
Did I just...

Yep, I edited you comment instead of writing a new one.

Great...let me fix it 🤦

**Comment by TrianguloY** [comment]
> There is also this graphical bug, I don't know why it happens, probably some basic knowledge about android I just lack, hopefully you are able to fix it. At the bottom, with the other buttons, the arrow of the spinner is just,... there, doing something. If you fix this, please tag me, I want to know why it happens.
> gfx_bug.mp4

@PabloOQ 

I found this: https://stackoverflow.com/questions/36167187/android-spinner-arrow-always-above-other-views

Seems to be an Android bug, but the solution suggested does work:
Just set `android:background="#00000000"` in the parent linearlayout of the spinner from the xml file (the post mentions the root parent, but it seems any parent works). Setting the layour type to software also works, but that's more dangerous to change I guess.

**Comment by PabloOQ** [comment]
> But as I said it's not easy to make it better. I'll try a few things, but if nothing works I'll probably just add a row header at the top and call it a day.

I see, well a row header is something I should definitely have added, that should be in there for sure.

> For example: the checkbox shown on the settings screen is visually the same as the one on the dialog one (it's the android standard after all, no problem there) but on one screen it's to enable/disable the flag, and on the other it's to show/hide it, while the dropdown is to enable/disable (more or less).

About the meaning of the checkbox, if these were to be visually different, it makes more sense to use the toggle in the settings and the checkbox in the main dialog, I think the main dialog should keep the checkbox. Some alternatives could be:

-  In the preferences, a spinner with show/hide, but it might a hassle when tweaking the preferences, however it makes very clear what it does and makes it completely different from the main dialog.
- In the preferences, a button which changes the inner text, although I usually find these totally confusing, is the text shown the current setting or is it what it will change to after pressing the button? With the spinner right next to it I think it is suggesting that the current text is the current setting, using "shown/hidden" as text would suggest the same too, instead of "show/hide"
- In the preferences a toggle with a text label next to it to clearly showing the current behaviour.

> Seems to be an Android bug, but the solution suggested does work:
Just set android:background="#00000000" in the parent linearlayout of the spinner from the xml file (the post mentions the root parent, but it seems any parent works). Setting the layour type to software also works, but that's more dangerous to change I guess.

Thanks! I spent quite a while trying to figure out what was the problem, but I didn't find anything. I'm glad it wasn't something on my end. I never thought this would be a bug so specific, I was looking for more general things, that might be why I couldn't find it. I hope it didn't took you long, I would feel awful.

**Comment by TrianguloY** [comment]
> Thanks! I spent quite a while trying to figure out what was the problem, but I didn't find anything. I'm glad it wasn't something on my end. I never thought this would be a bug so specific, I was looking for more general things, that might be why I couldn't find it. I hope it didn't took you long, I would feel awful.

No more than 5 minutes 😅
First I opened the layout editor from inside Android Studio, to see if there were something strange. I found nothing, but noticed that only the dropdown indicator was show above other views. So I just searched that: "android spinner dropdown shown above other views" and the first result was that link. I guess it was just 'easy' for me as an outsider, since you made the code you knew a lot of the details, so your search was more focused (which usually is a good thing, but not in some cases)

**Comment by PabloOQ** [comment]
Without any context I think the flag icon itself is easy to understand, at least the ones for enable and disable, as it follows the style of the checkbox, which is right next to it. I'm guessing subconsciously one would correlate that the tick and filled mean the same. With some context the crossed one can be deducted easily. I think using these for the dialog is a great idea.

Maybe using an icon for the checkbox could help too. An eye, crossed when the flag is hidden.

**Comment by TrianguloY** [comment]
image

image

It's taking me too long, but not because it's difficult, but because I haven't been able to spend more than two hours straight on this :/

**Comment by PabloOQ** [comment]
It looks great! I really like the dotted flag

**Comment by TrianguloY** [comment]
Too many changes :/

You can check them from the commits, but I still want to check a few things before merging.
In any case, can you check this version to see if I didn't break anything from your code? I hope not, but I changed a lot of things (probably more than I should have) so I want to be sure.

Hey! look at that, the new 'apk testing' works! Anyone can test now!

**Comment by PabloOQ** [comment]
Are you changing things on your local branch? There is a bug when there is no file (the eye is not attached to any listener), already fixed it my local branch. Aside from that everything seems fine.

> Hey! look at that, the new 'apk testing' works! Anyone can test now!

Pretty cool action!
Just letting you know, to check if this is intended behaviour. This can only be downloaded if logged in, anonymous users can't download the file. I also asked a friend to check if you need specials permissions to download it, as I could download it with my account, but I do have some permissions in the repo, he could download it.

**Comment by TrianguloY** [comment]
> Are you changing things on your local branch? There is a bug when there is no file (the eye is not attached to any listener), already fixed it my local branch. Aside from that everything seems fine.

I detected it, but haven't been able to fix it yet. Upload your fix please (if possible)

> Pretty cool action!
> Just letting you know, to check if this is intended behaviour. This can only be downloaded if logged in, anonymous users can't download the file. I also asked a friend to check if you need specials permissions to download it, as I could download it with my account, but I do have some permissions in the repo, he could download it. 

Oh, I though it was available publicly, thanks for mentioning. 
Requiring a Github account is not bad (this is intended for people publishing translations mainly, and those always have an account to create the pr in the first place) but I'll try to see if it can be changed.

**Comment by TrianguloY** [comment]
Merged! It can be fixed later if we discover an issue in any case.
